File: erpnext/gp_erp/report/budget_variance_greenphyto/budget_variance_greenphyto.py
import frappe
from frappe import _
from frappe.utils import flt, get_first_day, get_last_day, getdate


def execute(filters=None):
	filters = control_filters(filters)
	columns = get_columns(filters)
	data = get_data(filters)
	return columns, data


def control_filters(filters):
	if not filters:
		filters = frappe._dict()

	year = filters.get("year")
	month = filters.get("month")
	to_month = filters.get("to_month")

	if year:
		filters.from_fiscal_year = year
		filters.to_fiscal_year = year

	if year and month:
		date_str = f"{year}-{month}-01"
		filters.period_start_date = get_first_day(date_str)

	if year and to_month:
		to_date_str = f"{year}-{to_month}-01"
		filters.period_end_date = get_last_day(to_date_str)

	return filters


def get_columns(filters):
	return [
		{"fieldname": "account", "label": _("Account"), "fieldtype": "Link", "options": "Account", "width": 300},
		{"fieldname": "total_actual", "label": _("Total Actual"), "fieldtype": "Currency", "width": 150},
		{"fieldname": "budget_ytd", "label": _("Budget YTD"), "fieldtype": "Currency", "width": 150},
		{"fieldname": "variance_amount", "label": _("Variance"), "fieldtype": "Currency", "width": 150},
		{"fieldname": "variance_percent", "label": _("Variance %"), "fieldtype": "Percent", "width": 100},
	]


def get_data(filters):
	budget_accounts = get_budget_account(
		cost_center=filters.get("cost_center"),
		company=filters.get("company"),
	)
	budget_data = get_budget_data(filters)
	return budget_accounts


def get_budget_account(cost_center=None, company=None):
	conditions = {"company": company} if company else {}

	if cost_center:
		budgets = frappe.db.get_all("Budget", filters={
			"docstatus": 1,
			"cost_center": ["in", cost_center],
			**conditions,
		}, pluck="name")
	else:
		budgets = frappe.db.get_all("Budget", filters={
			"docstatus": 1,
			**conditions,
		}, pluck="name")

	accounts = []
	for budget in budgets:
		items = frappe.db.get_all("Budget Account", filters={"parent": budget}, fields=["account", "budget_amount"])
		for item in items:
			accounts.append(item)

	return accounts


def get_budget_data(filters):
	data = {}

	company = filters.get("company")
	from_fiscal_year = filters.get("from_fiscal_year")
	to_fiscal_year = filters.get("to_fiscal_year")
	budget_against = filters.get("budget_against", "Cost Center")

	if not company or not from_fiscal_year:
		return data

	fy_filters = {"docstatus": 1, "company": company}
	if from_fiscal_year and to_fiscal_year and from_fiscal_year != to_fiscal_year:
		fiscal_years = frappe.db.get_all("Fiscal Year",
			filters={"name": [">=", from_fiscal_year]},
			pluck="name"
		)
		fiscal_years = [fy for fy in fiscal_years if fy <= to_fiscal_year]
		if not fiscal_years:
			fiscal_years = [from_fiscal_year]
		fy_filters["fiscal_year"] = ["in", fiscal_years]
	else:
		fy_filters["fiscal_year"] = from_fiscal_year

	budgets = frappe.db.get_all("Budget", filters=fy_filters,
		fields=["name", "cost_center", "fiscal_year"])

	for budget in budgets:
		items = frappe.db.get_all("Budget Account", filters={"parent": budget.name}, fields=["account", "budget_amount"])
		for item in items:
			key = (budget.cost_center, item.account)
			if key not in data:
				data[key] = {"budget_amount": 0}
			data[key]["budget_amount"] += flt(item.budget_amount)

	return data


def add_summary_columns(rows, period_list):
	for row in rows:
		if row.get("profit_data"):
			continue

		total_actual = 0
		budget_ytd = 0

		for period in period_list:
			key = period.get("key")
			total_actual += flt(row.get(key, 0))
			budget_ytd += flt(row.get(f"{key}_budget", 0))

		row["total_actual"] = total_actual
		row["budget_ytd"] = budget_ytd
		row["variance_amount"] = total_actual - budget_ytd

		if budget_ytd:
			row["variance_percent"] = flt((total_actual - budget_ytd) / budget_ytd * 100, 1)
		else:
			row["variance_percent"] = 0

File: erpnext/gp_erp/report/budget_variance_greenphyto/budget_variance_greenphyto.js
// Copyright (c) 2015, Frappe Technologies Pvt. Ltd. and Contributors
// License: GNU General Public License v3. See license.txt
frappe.provide("erpnext.financial_statements");

erpnext.financial_statements = {
	"filters": get_filters(),
	"formatter": function(value, row, column, data, default_formatter) {
		if (data && column.fieldname=="account") {
			value = data.account_name || value;

			column.link_onclick =
				"erpnext.financial_statements.open_general_ledger(" + JSON.stringify(data) + ")";
			column.is_tree = true;
		}

		value = default_formatter(value, row, column, data);

		if (data && data.is_group) {
			value = $(`<span>${value}</span>`);

			var $value = $(value).css("font-weight", "bold");
			if (data.warn_if_negative && data[column.fieldname] < 0) {
				$value.addClass("text-danger");
			}

			value = $value.wrap("<p></p>").parent().html();
		}

		return value;
	},
	"open_general_ledger": function(data) {
		if (!data.account) return;
		let project = $.grep(frappe.query_report.filters, function(e){ return e.df.fieldname == 'project'; });

		frappe.route_options = {
			"account": data.account,
			"company": frappe.query_report.get_filter_value('company'),
			"from_date": data.from_date || data.year_start_date,
			"to_date": data.to_date || data.year_end_date,
			"project": (project && project.length > 0) ? project[0].$input.val() : ""
		};

		let report = "General Ledger";

		if (["Payable", "Receivable"].includes(data.account_type)) {
			report = data.account_type == "Payable" ? "Accounts Payable" : "Accounts Receivable";
			frappe.route_options["party_account"] = data.account;
			frappe.route_options["report_date"] = data.year_end_date;
		}

		frappe.set_route("query-report", report);
	},
	"tree": true,
	"name_field": "account",
	"parent_field": "parent_account",
	"initial_depth": 3,
	onload: function(report) {
		// dropdown for links to other financial statements
		erpnext.financial_statements.filters = get_filters()

		let fiscal_year = frappe.defaults.get_user_default("fiscal_year")

		frappe.model.with_doc("Fiscal Year", fiscal_year, function(r) {
			var fy = frappe.model.get_doc("Fiscal Year", fiscal_year);
			frappe.query_report.set_filter_value({
				period_start_date: fy.year_start_date,
				period_end_date: fy.year_end_date
			});
		});

		const views_menu = report.page.add_custom_button_group(__('Financial Statements'));

		report.page.add_custom_menu_item(views_menu, __("Balance Sheet"), function() {
			var filters = report.get_values();
			frappe.set_route('query-report', 'Balance Sheet', {company: filters.company});
		});

		report.page.add_custom_menu_item(views_menu, __("Profit and Loss"), function() {
			var filters = report.get_values();
			frappe.set_route('query-report', 'Profit and Loss Statement', {company: filters.company});
		});

		report.page.add_custom_menu_item(views_menu, __("Cash Flow Statement"), function() {
			var filters = report.get_values();
			frappe.set_route('query-report', 'Cash Flow', {company: filters.company});
		});
	}
};

function get_filters() {
	let filters = [
		{
			"fieldname":"company",
			"label": __("Company"),
			"fieldtype": "Link",
			"options": "Company",
			"default": frappe.defaults.get_user_default("Company"),
			"reqd": 1
		},
		{
			"fieldname":"year",
			"label": __("Year"),
			"fieldtype": "Link",
			"options": "Fiscal Year",
			"default": frappe.defaults.get_user_default("fiscal_year"),
			"reqd": 1
		},
		{
			"fieldname":"month",
			"label": __("From Month"),
			"fieldtype": "Select",
			"reqd": 0,
			"options": "\nJanuary\nFebruary\nMarch\nApril\nMay\nJune\nJuly\nAugust\nSeptember\nOctober\nNovember\nDecember",
		},
		{
			"fieldname":"to_month",
			"label": __("To Month"),
			"fieldtype": "Select",
			"reqd": 0,
			"options": "\nJanuary\nFebruary\nMarch\nApril\nMay\nJune\nJuly\nAugust\nSeptember\nOctober\nNovember\nDecember",
		},
		{
			"fieldname": "periodicity",
			"label": __("Periodicity"),
			"fieldtype": "Select",
			"options": [
				{ "value": "Monthly", "label": __("Monthly") },
				{ "value": "Yearly", "label": __("Yearly") },
			],
			"default": "Yearly",
			"reqd": 1
		},
		// Note:
		// If you are modifying this array such that the presentation_currency object
		// is no longer the last object, please make adjustments in cash_flow.js
		// accordingly.
		{
			"fieldname": "presentation_currency",
			"label": __("Currency"),
			"fieldtype": "Select",
			"options": erpnext.get_presentation_currency_list()
		},
		{
			"fieldname": "cost_center",
			"label": __("Cost Center"),
			"fieldtype": "MultiSelectList",
			get_data: function(txt) {
				return frappe.db.get_link_options('Cost Center', txt, {
					company: frappe.query_report.get_filter_value("company")
				});
			}
		}
	]

	return filters;
}



frappe.query_reports["Budget Variance Greenphyto"] = $.extend({},
	erpnext.financial_statements);

erpnext.utils.add_dimensions('Budget Variance Greenphyto', 10);

frappe.query_reports["Budget Variance Greenphyto"]["filters"].push(
	{
		"fieldname": "accumulated_values",
		"label": __("Accumulated Values"),
		"fieldtype": "Check",
		"default": 1
	},
	{
		"fieldname": "hide_zero_balance",
		"label": __("Hide Zero Balance"),
		"fieldtype": "Check",
		"default": 0
	}
);

frappe.query_reports["Budget Variance Greenphyto"]["onload"] = function(report){
	report.page.add_inner_button("Export with Cost Centers", function() {
		frappe.call({
			method: "erpnext.gp_erp.report.budget_variance_greenphyto.budget_variance_greenphyto.get_export_with_cost_centers_url",
			args: {
				filters: report.get_values()
			},
			callback: function(r) {
				if (r.message && r.message.url) {
					window.open(r.message.url);
				}
			}
		});
	});
}
